### ISSUE TITLE
CMake build issues

### ISSUE THREAD

**Issue by mihaipopescu** [uncrustify/uncrustify#604]
@jibsen I've just tested cmake build on windows and apparently it works but:
- [x] there are no instructions on how to build with cmake
- [x] cmake fails when it tries to create a `/Testing` directory because there is another file with the same name `/TESTING`
- [x] seems like process is rather messy. It creates everything in the root and pollutes the repository with untracked files.
- [x] modify `appveyor.yml` and `.travis.yml` to switch to cmake
- [x] Add appveyor build status badge in README.md
- [x] add helper script for running tests
- [x] make_version.py is not working
- [ ] repository has to be cleaned of old build system files (gnu automake, mingw32 scripts, VS and Xcode solution files)

**Comment by mihaipopescu** [comment]
Other related issues #470, #538, #600.

**Comment by jibsen** [comment]
I would always recommend doing an out-of-source build with CMake, so all the files created do not end up in the source tree:

~~~.cmd
mkdir build
cd build
cmake ..
~~~

Cleaning up is just a matter of removing the build folder.

**Comment by mihaipopescu** [comment]
Can't the output/temp folders be specified ? I would be very surprised if not.

**Comment by jibsen** [comment]
I may be wrong, but I think CMake always creates them in the folder cmake is run from. Perhaps it would be worth adding a check to the CMake script that warns the user if attempting an in-source build?

Regarding cleaning up the other build systems, I think it is up to the authors if they feel comfortable enough with CMake as the only option.

**Comment by mihaipopescu** [comment]
I have no problem with changing the `cwd` if that's the only way for cmake. I have little to none experience with CMake.

I would agree with any build system as long as it's cross platform and can create native executables for that OS.

**Comment by mihaipopescu** [comment]
@jibsen I've branched the discussion into a new issue #606. We can keep this one only for CMake related issues.

**Comment by jibsen** [comment]
Thanks. I've opened #605 and #608 to address some of these.

**Comment by mihaipopescu** [comment]
Thanks @jibsen. I'll try them out.

**Comment by mihaipopescu** [comment]
Let me know if you think there should be something else.

**Comment by jibsen** [comment]
The output files from the tests should probably go into the build folder somewhere (using the `--results` option). I'll take a look at that.

**Comment by mihaipopescu** [comment]
Why ? I think it's good where they are right now in the `tests` folder. The test fw just has to find the executable, that's all I guess.

**Comment by jibsen** [comment]
It would keep generated files out of the source tree, so cleaning up is just removing the build folder. Also, for testing, I have different build folders for mingw-w64 and MSVC, and this would keep the results with the corresponding build.

But perhaps I am overcomplicating things?

**Comment by mihaipopescu** [comment]
What if we have a root `CMakeLists.txt` which just includes a `tests/CMakeLists.txt` which only knows about tests, and how to clean them etc. and a `src/CMakeLists.txt` which knows how to build uncrustify. And there is also `man` folder which I'm unsure of how `cmake` will be able to make it (probably only on UNIX).

With the `--results` option yeah, the result folder can be moved elsewhere at user's preference and you can make it per distribution. I wasn't sure if anyone was still using mingw build.

**Comment by jibsen** [comment]
Splitting up the CMake script is probably a good idea in the long run.

It looks like the man page is just a question of running something like:

~~~.cmake
configure_file(man/uncrustify.1.in uncrustify.1 @ONLY)
~~~

**Comment by mihaipopescu** [comment]
> Splitting up the CMake script is probably a good idea in the long run.

I agree with that.

**Comment by jibsen** [comment]
As mentioned in the other thread, perhaps break up the tests in the CMake script so there is one for each language, making it possible to run them individually.

**Comment by mihaipopescu** [comment]
I think in general it would make sense that `ctest` would have as many options from `run_tests.py` including language and test filtering for instance.

Another thing I've noticed is that it looks unresponsive until tests finish and I'm not sure why.

**Comment by mihaipopescu** [comment]
We also need @dirk-thomas's and @gmaurel to add AppVeyor to uncrustify.

@dirk-thomas You need to make the appveyor PR https://github.com/dirk-thomas/uncrustify/pull/1 against upstream, not to your fork.

**Comment by jibsen** [comment]
#610 splits up the CTest tests by language.

> Another thing I've noticed is that it looks unresponsive until tests finish and I'm not sure why.

CTest lets the entire test run, capturing the output (without the `-V` option, it only prints if the test succeeded (returned zero), not the output from it). With one monolithic test that means you do not see any result until `run_tests.py` is finished.

**Comment by jibsen** [comment]
#612 switches to CMake for the Travis build. I added an address sanitizer build as well, not sure if that's something you want though.

**Comment by dirk-thomas** [comment]
@mihaipopescu The AppVeyor project should be created before the PR (as mentioned [here](https://github.com/uncrustify/uncrustify/issues/538#issuecomment-233959610)).

**Comment by mihaipopescu** [comment]
@gmaurel can you do it ?

**Comment by mihaipopescu** [comment]
@jibsen I've noticed a consistent drop of the Release executable size of the build created with CMake (with around 100kb from 500kb). Are there any differences in optimization flags between CMake and old Visual Studio build (which I guess was optimized for speed) ?

**Comment by mihaipopescu** [comment]
Also, does CMake detect when `token_enum.h` is modified to trigger a rebuild of `token_names.h` ?

**Comment by jibsen** [comment]
> I've noticed a consistent drop of the Release executable size of the build created with CMake (with around 100kb from 500kb)

Judging by the size difference, perhaps the old build was linking with static runtime libraries?

> Also, does CMake detect when token_enum.h is modified to trigger a rebuild of token_names.h ?

It actually looks like it doesn't. I'll have a look, thanks.

**Comment by mihaipopescu** [comment]
Yes, that seems to have been the difference. Hum, that means that users need to have the runtime redistributable installed on their system as a prerequisite.

Also, another partially-related issue is that the `DEBUG` symbol needs to be defined on `Debug` configuration (maybe also on `RelWithDebInfo`). This enables callstacks to be generated for log functions.

**Comment by jibsen** [comment]
> Also, does CMake detect when token_enum.h is modified to trigger a rebuild of token_names.h ?

#618

> Yes, that seems to have been the difference. Hum, that means that users need to have the runtime redistributable installed on their system as a prerequisite.

It is possible to [force a static link](https://cmake.org/Wiki/CMake_FAQ#How_can_I_build_my_MSVC_application_with_a_static_runtime.3F) in a couple of ways. I am leaning towards the dynamic replace, since it's simple, and this project is only building one executable target.

> Also, another partially-related issue is that the DEBUG symbol needs to be defined on Debug configuration (maybe also on RelWithDebInfo). This enables callstacks to be generated for log functions.

I think the easiest fix would be to check if `NDEBUG` is not defined instead, which is set for all builds but Debug. Of course it's possible to add a DEBUG flag, there are just some strange cases like Makefiles having no build type by default.

**Comment by mihaipopescu** [comment]
> It is possible to force a static link in a couple of ways. I am leaning towards the dynamic replace, since it's simple, and this project is only building one executable target.

Let's see about that. I'm really in favor of having no dependency but nowadays that's almost impossible with VS. Every system ends up having literally dozens of runtime redistributables from 2005 to 2015.

> I think the easiest fix would be to check if NDEBUG is not defined instead, which is set for all builds but Debug. Of course it's possible to add a DEBUG flag, there are just some strange cases like Makefiles having no build type by default.

`NDEBUG` is an option as well but it's mostly a VS convention as well as `_DEBUG`. Check also `uncrustify_version.h` and the rest of the code for usage of `DEBUG` defines and let's just settle on one.

**Comment by mihaipopescu** [comment]
And now that I think about it I think we should have used a branch for cmake feature and then PR that back into upstream.

**Comment by jibsen** [comment]
> NDEBUG is an option as well but it's mostly a VS convention as well as _DEBUG.

Actually `NDEBUG` is part of the C standard, it controls whether `assert()` does anything. As I understand it, MSVC uses `_DEBUG` to control debugging features in the runtime.

I guess it depends on if the debugging enabled by `DEBUG` in Uncrustify should be linked to either of those, or toggled independently?

**Comment by mihaipopescu** [comment]
> Actually NDEBUG is part of the C standard
Now I understand my confusion and why it didn't worked my previous implementation https://github.com/Unity-Technologies/uncrustify/blob/53679f44dda5d6579cfaee82d56133571d7db2f5/src/uncrustify_version.h.in. I mean, it worked on VS builds but not on UNIX builds where I thought it should have been defined through the autotools or at least in the Xcode project.

> I guess it depends on if the debugging enabled by DEBUG in Uncrustify should be linked to either of those, or toggled independently?

I think we should have a way to control `DEBUG` define which controls verbose logs with callstacks. We need this in Debug builds and so it probably should be toggled depending the value of `NDEBUG`.

```cpp
#if !defined(NDEBUG)
#define DEBUG
#endif
```

or defined independently with CMake which probably it's better.

It's clear that we should not touch `_DEBUG` which is a specific MSVC define.

**Comment by jibsen** [comment]
> or defined independently with CMake which probably it's better.

I tried adding that in #619.

**Comment by mihaipopescu** [comment]
`master` branch started using cmake https://travis-ci.org/uncrustify/uncrustify/builds. Let's see how it goes.

![matrix](https://cloud.githubusercontent.com/assets/272470/17522984/56d67fcc-5e59-11e6-9682-46cff3ade176.JPG)

I'm a bit concerned about that now the build takes much more time. Maybe we can reduce the scope of the build like build just release.

**Comment by dirk-thomas** [comment]
Since these different builds should happen in parallel they shouldn't increase the time noticeably.

**Comment by mihaipopescu** [comment]
ok but even with that we've increased from 1 to 10 minutes.

**Comment by dirk-thomas** [comment]
The average build seems to be between 2 and 4 minutes. (One build took 8 minutes but that was an exception as far as I can see.)

**Comment by mihaipopescu** [comment]
Yeah, all right. It's still a 3x increase but I guess it's a reasonable price to pay.

**Comment by mihaipopescu** [comment]
I like and appreciate the effort you guys are putting into this. There are few issues remaining until the transition is complete (more or less).

When I added
> add helper script for running tests

I was thinking to have some sort of wrapper over `run_tests.py` to be take advantage of all it's arguments (it's slightly related with #600).

The thing with `ctest` seemed quite limited with few command line options. I was looking for something similar to double hyphen `--` positional argument which would separate normal arguments to the ones that are passed to `run_tests.py`.

```shell
ctest -V -C Debug -- cpp -r 33000-33010 -d -g
```

This would basically append the correct `--exe` argument but pass the other arguments after `-- ` as they are.

On the other hand, from what I understand is that everything would be just one test. But the thing is we can't have a complete cmake test feature if we don't parse `.test` files the same way. But why do it when we already have it in python. This could be the price we have to pay if we want to link the two testing frameworks.

What do you guys think ?

**Comment by mihaipopescu** [comment]
Conversely, another way would be if we could have a way to return the executable name from cmake giving it a configuration and stick using the `run_tests` scripts. I think that was the idea from #600.

**Comment by dirk-thomas** [comment]
Registering a single script with `add_test` and running all the test within it significantly reduces the usefulness of `ctest`. `ctest` has so many useful options (https://cmake.org/cmake/help/v3.5/manual/ctest.1.html) and most of them couldn't be applied in that case.

I would suggest to register each test separately in CMake by calling `add_test`. That e.g. allows us to give them labels, the path to the executable is not an issue anymore since it can be resolved by CMake using a generator expression, etc. Each test should probably still be wrapped in a small script which continues to bundle the logic currently implemented in the `run_tests` function (https://github.com/uncrustify/uncrustify/blob/5bdecf49d9d1f04cc5806871697b63d50df3a2a3/tests/run_tests.py#L86) (run, compare output, rerun, compare again).

If people agree that this would be a good approach I can try to look into it.

**Comment by mihaipopescu** [comment]
Sure, porting `run_tests.py` to cmake would be ideal but it involves a lot of work. If you are willing to start working on it would be awesome.

**Comment by dirk-thomas** [comment]
I will take a look once we have CI for both platforms. That way I can actually ensure to not break anything or miss any tests :wink:

**Comment by dirk-thomas** [comment]
Please see #627.

**Comment by mihaipopescu** [comment]
@dirk-thomas That change actually introduces a lot of overhead launching python interpreter for each test and heavy command line logging. Even if you run it in parallel it's slow compared to the old python script.

**Comment by dirk-thomas** [comment]
The logging is not fixed. You can configure it to your liking. If you run `ctest -Q` it won't output anything at all.

While the Python invocations imply some overhead the tests are by no means slower. I can run all tests in less then 5s on a notebook with no output (`ctest -j 16 -Q`) With the `run_tests.py` script it takes 8s. So I would say that is quite the opposite.

**Comment by dzenanz** [comment]
I was surprised to see CMake build system and a custom solution for regression tests. Porting it to ordinary CMake tests would drop the dependence on Python which is not frequently installed on Windows. So @dirk-thomas , you have my vote for implementing the switch!

**Comment by dirk-thomas** [comment]
@dzenanz As mentioned in [this comment](https://github.com/uncrustify/uncrustify/issues/604#issuecomment-239186059) we still need Python since each test beside calling uncrustify needs to compare the result and optionally rerun the formatter. That logic needs to be encapsulated somewhere (and I would assume writing a bash and batch script would be much more painful then the Python dependency).

**Comment by dzenanz** [comment]
Even without dropping the Python dependency it would be worthwhile to have a meaningful output by invoking the tests in a standard way (ctest from command line or RUN_TESTS from Visual Studio).

**Comment by dirk-thomas** [comment]
Indeed. I especially like the CTest option `--rerun-failed`.

**Comment by mihaipopescu** [comment]
I don't think the argument that you can run many tests in parallel can counter the fact that the testing framework has been significantly slowed down for single thread usage. Parallelization can be achieved even with previous framework simply by running multiple commands in parallel: `python run_tests.py c & python run_tests.py cpp & ...`. Counting also the fact that uncrustify has also the batch option even faster testing can be achieved.

**Comment by mihaipopescu** [comment]
We need to port more functionality from `run_scripts.py` to call that script redundant. Till then we need it.

Still, I've found a new issue identified while running the `make_version.py` which doesn't work with cmake. We have to figure it out how to fix this but I would just call it from cmake.

**Comment by jibsen** [comment]
> Still, I've found a new issue identified while running the `make_version.py` which doesn't work with cmake. We have to figure it out how to fix this but I would just call it from cmake.

Since you cannot necessarily count on people building Uncrustify having cloned the git repository, I don't think we can do what `make_version.py` does (`git describe`) as part of the regular build process.

We could put `set(PACKAGE_VERSION "0.63")` in `CMakeLists.txt` and then update that in the script, or we could write a version tag file and read that from the CMake script?

**Comment by mihaipopescu** [comment]
@jibsen `make_version.py` is doing some checks to see if there is a `.git` or an `.hg` repository and then runs respective commands to find the commit hashes. I won't be a good idea though to make it part of the build system, that's not what I'm saying. We just have to make it work with cmake.

> We could put set(PACKAGE_VERSION "0.63") in CMakeLists.txt and then update that in the Python script, or we could write a version tag file and read that from the CMake script?

Agree. We have to clean the dependency on `configure.ac` dependency. Also we can send the build directory as a parameter to be used as an output directory instead of the src folder.

What do you think ?

**Comment by mihaipopescu** [comment]
We also have a fix for users that are using [hg-git](http://hg-git.github.io/) extension.

```diff
--- C:\Users\mihai\AppData\Local\Temp\thg._cjm9b\uncrustify.6.1765\make_version.py	2016-08-26 15:32:51.000000000 +0200
+++ C:\Users\mihai\AppData\Local\Temp\thg._cjm9b\uncrustify.6\make_version.py	2016-08-26 15:32:51.000000000 +0200
@@ -16,14 +16,15 @@
             txt = subprocess.check_output(['git', 'describe', '--long', '--always', '--dirty']).strip()
         except:
             print "Failed to retrieve version from git"
             return 1
     elif os.path.exists(hg_path):
         try:
-            branch = subprocess.check_output(['hg', 'log', '-r', '.', '--template', '{latesttag}']).strip()
-            txt = subprocess.check_output(['git', '--git-dir=.hg/git', 'describe', '--long', '--always', branch]).strip()
+            subprocess.check_call(['hg', 'gexport'])
+            node = subprocess.check_output(['hg', '--config', 'defaults.log=', 'log', '-r', '.', '--template', '{gitnode}']).strip()
+            txt = subprocess.check_output(['git', '--git-dir=.hg/git', 'describe', '--long', '--tags', '--always', node]).strip()
         except:
             print "Failed to retrieve version from hg"
             return 1
     else:
         print "Unknown version control system."
         return 1
```

We can add this to the patch.

**Comment by mihaipopescu** [comment]
I've made a patch for the last issue I've noted before doing the full rollout to cmake.
Do you guys have anything else to add ? I may have missed something.

**Comment by gmaurel** [comment]
Do we have anything else to do?
May I close?

**Comment by dirk-thomas** [comment]
I think the last bullet from the first comment is still pending (repository has to be cleaned of old build system files).

**Comment by gmaurel** [comment]
Which one may I remove?

**Comment by dirk-thomas** [comment]
I haven't tried it but from looking at the file system I would guess the following should not be used anymore:

* autotools/
* uncrustify.xcodeproj/
* win32/
* Makefile.am
* autogen.sh
* configure.ac
* make_win32.sh

It looks like the following files need to be update to use the new build system / remove / replace autotools related stuff:

* INSTALL
* README.md
* TESTING
* package.json
* release-steps.txt

**Comment by dzenanz** [comment]
What about folders `man` and `osx`, and files `Comments.txt`, `extras.vpj`, `uncrustify.vpj`, `uncrustify.vpw` and `working.txt`?

**Comment by mihaipopescu** [comment]
It's a good time for cleanup. I think we should create two different PRs. One for the old build system cleanup and one for general cleanup.

```shell
# gnu autotools cleanup
rm -rf autotools/
rm -rf uncrustify.xcodeproj/
rm -rf win32/
rm -rf Makefile.am
rm autogen.sh
rm configure.ac
rm make_win32.sh
rm make_token_names.*
mv make_version scripts/
rm src/Makefile.am
```

```shell
# general cleanup
rm -rf osx
rm commit.log
rm extras.vpj
mv forUncrustifySources.* etc/
rm uncrustify.vp*
mv working.txt documentation/
rm NEWS
```

I think `man` it's still used from cmake to generate the manual. We have to check if all files are required.

**Comment by jibsen** [comment]
I started a PR for cleaning up the build system files.

We should probably also resolve #650 before any new release.

**Comment by mihaipopescu** [comment]
I'm going to close this now as very few lingering issues remained.
